packages/@react-stately/table/src/TableUtils.ts

```diff
@@ -133,42 +65,194 @@ export interface IColumn {
+ * @param getDefaultWidth - A function that returns the default width of a column by its index.
+ * @param getDefaultMinWidth - A function that returns the default min width of a column by its index.
+ */
 export function calculateColumnSizes(availableWidth: number, columns: IColumn[], changedColumns: Map<Key, ColumnSize>, getDefaultWidth, getDefaultMinWidth) {
-  let remainingSpace = availableWidth;
-  let {staticColumns, dynamicColumns} = columns.reduce((acc, column, index) => {
+  let initialUsedSpace = 0;
+  let hasNonFrozenItems = false;
+  let flexItems = columns.map((column, index) => {
     let width = changedColumns.get(column.key) != null ? changedColumns.get(column.key) : column.width ?? column.defaultWidth ?? getDefaultWidth?.(index) ?? '1fr';
-    let minWidth = column.minWidth ?? getDefaultMinWidth?.(index);
-    column.minWidth = minWidth;
-
+    let frozen = false;
+    let baseSize = 0;
+    let flex = 0;
+    let targetMainSize = null;
     if (isStatic(width)) {
-      let w = parseStaticWidth(width, availableWidth);
-      w = Math.max(
-        getMinWidth(column.minWidth, availableWidth),
-        Math.min(Math.floor(w), getMaxWidth(column.maxWidth, availableWidth)));
-      acc.staticColumns.push({index, column, width: w} as IndexedColumn);
-      acc.dynamicColumns.push(null);
-      remainingSpace -= w;
+      baseSize = parseStaticWidth(width, availableWidth);
+      frozen = true;
     } else {
-      let w = parseFractionalUnit(width);
-      acc.staticColumns.push(null);
-      acc.dynamicColumns.push({index, column, width: w} as IndexedColumn);
+      flex = parseFractionalUnit(width);
+      if (flex <= 0) {
+        frozen = true;
+      }
+    }
+
+    let min = getMinWidth(column.minWidth ?? getDefaultMinWidth?.(index) ?? 0, availableWidth);
+    let max = getMaxWidth(column.maxWidth, availableWidth);
+    let hypotheticalMainSize = Math.max(min, Math.min(baseSize, max));
+
+    // 9.7.1
+    // We don't make use of flex basis, it's always 0, so we are always in 'grow' mode.
+    // 9.7.2
+    if (frozen) {
+      targetMainSize = hypotheticalMainSize;
+    } else if (baseSize > hypotheticalMainSize) {
+      frozen = true;
+      targetMainSize = hypotheticalMainSize;
     }
-    return acc;
-  }, {staticColumns: [] as IndexedColumn[], dynamicColumns: [] as IndexedColumn[]});
-  let newColWidths = getDynamicColumnWidths(dynamicColumns, remainingSpace, availableWidth);
 
-  return staticColumns.map((col, i) => {
-    if (col) {
-      return col.width;
+    // 9.7.3
+    if (frozen) {
+      initialUsedSpace += targetMainSize;
+    } else {
+      initialUsedSpace += baseSize;
+      hasNonFrozenItems = true;
     }
-    return newColWidths[i].width;
+    return {
+      frozen,
+      baseSize,
+      hypotheticalMainSize,
+      min,
+      max,
+      flex,
+      targetMainSize,
+      violation: 0
+    };
   });
+
+  let remainingFreeSpace = availableWidth - initialUsedSpace;
```

Comment:
Nit: We can get rid this now and just initialize this var inside the while loop. I found it a bit confusing that we calculated this here but then completely recalculate it inside the while loop